### ISSUE TITLE
自定义模型想使用generate函数有哪些需要注意的地方？

### ISSUE THREAD

**Issue by Wu-Jiayang** [PaddlePaddle/PaddleNLP#1595]
- 版本、环境信息
1）PaddleNLP和PaddlePaddle版本：PaddleNLP 2.2.3，PaddlePaddle2.2.1

刚接触paddle的本科生小白，想修改模型结构，请教一下，自己搭建的模型想使用generate函数，除了需要继承GenerationMixin类、实现forward函数以外，还需要实现或重写哪些函数？或者说有哪些需要注意的地方？比如多输出模型？

**Comment by chenxiaozeng** [comment]
你好，如果是常见的预训练模型，PaddleNLP都是内置好了的。可以描述下具体需求，比如是想复现哪篇paper吗？

**Comment by smallv0221** [comment]
Usually it is better to Inherit `PretrainedModel` class instead of `GenerationMixin`. And you may need to write something like `GPTLMHeadModel` which contains your own `prepare_inputs_for_generation()` method.

**Comment by Wu-Jiayang** [comment]
> 你好，如果是常见的预训练模型，PaddleNLP都是内置好了的。可以描述下具体需求，比如是想复现哪篇paper吗？

自己写paper

**Comment by Wu-Jiayang** [comment]
> Usually it is better to Inherit `PretrainedModel` class instead of `GenerationMixin`. And you may need to write something like `GPTLMHeadModel` which contains your own `prepare_inputs_for_generation()` method.

I plan to inherit the `UnifiedTransformerPretrainedModel` class and write something like `UnifiedTransformerModel` and `UnifiedTransformerLMHeadModel`. Is `prepare_inputs_for_generation()` method the only thing I need to concern if I write a multiple-input and multiple-output model?
![image](https://user-images.githubusercontent.com/62842862/149903866-47a237df-0c23-4dad-a7a5-bcac6d0807d1.png)

**Comment by Wu-Jiayang** [comment]
> 你好，如果是常见的预训练模型，PaddleNLP都是内置好了的。可以描述下具体需求，比如是想复现哪篇paper吗？

想在传统UnifiedTransformerLMHeadModel模型的基础上增加输入参数和输出参数，并使用generate函数进行生成

**Comment by smallv0221** [comment]
As you can see in `generation_utils.py`.
![image](https://user-images.githubusercontent.com/33639025/149910047-dd4488f9-f479-4194-badf-b41138ab6102.png)
There are only two main concern.
1. Make sure the output of your `prepare_inputs_for_generation` and the input of your Model matches.
2. If you Model have multiple-output, It must return `Tuple` and `logits` must be the first element of that tuple.

**Comment by Wu-Jiayang** [comment]
> As you can see in `generation_utils.py`. ![image](https://user-images.githubusercontent.com/33639025/149910047-dd4488f9-f479-4194-badf-b41138ab6102.png) There are only two main concern.
> 
> 1. Make sure the output of your `prepare_inputs_for_generation` and the input of your Model matches.
> 2. If you Model have multiple-output, It must return `Tuple` and `logits` must be the first element of that tuple.

Thanks a lot ! I will have a try.

**Comment by Wu-Jiayang** [comment]
> As you can see in `generation_utils.py`. ![image](https://user-images.githubusercontent.com/33639025/149910047-dd4488f9-f479-4194-badf-b41138ab6102.png) There are only two main concern.
> 
> 1. Make sure the output of your `prepare_inputs_for_generation` and the input of your Model matches.
> 2. If you Model have multiple-output, It must return `Tuple` and `logits` must be the first element of that tuple.

@smallv0221 I found that my second output is used as cache somewhere and that leads to some mistakes, how can I fix this?

**Comment by smallv0221** [comment]
![image](https://user-images.githubusercontent.com/33639025/150492711-73d3ae4d-e12f-4ed3-9744-b2d03e3acd0a.png)
If the second output is not cache, you may need to rewrite `update_model_kwargs_for_generation` as well.

**Comment by Wu-Jiayang** [comment]
> ![image](https://user-images.githubusercontent.com/33639025/150492711-73d3ae4d-e12f-4ed3-9744-b2d03e3acd0a.png) If the second output is not cache, you may need to rewrite `update_model_kwargs_for_generation` as well.

Thanks a lot!
